File: 2140-solving-questions-with-brainpower/2140-solving-questions-with-brainpower.py
class Solution:
    def mostPoints(self, questions: List[List[int]]) -> int:
        N = len(questions)
        dp = [0] * N
        dp[-1] = questions[-1][0]
        for i in range(N - 2, -1, -1):
            next = i + questions[i][1] + 1             
            dp[i] = max(dp[i + 1], questions[i][0] + (dp[next] if next < N else 0))
        return dp[0]

File: 2140-solving-questions-with-brainpower/2140-solving-questions-with-brainpower.java
class Solution {
    Long[] dp;
    public long mostPoints(int[][] questions) {
        dp = new Long[questions.length];
        return dfs(questions, 0);
    }

    private long dfs(int[][] questions, int i) {
        if (i >= questions.length) return 0;
        if (dp[i] != null) return dp[i];
        long pick = questions[i][0] + dfs(questions, i + questions[i][1] + 1);
        long skip = dfs(questions, i + 1);
        return dp[i] = Math.max(pick, skip);
    }
}